### ISSUE TITLE
processing messages without keys

### ISSUE THREAD

**Issue by mikedewar** [lovoo/goka#374]
Hi!

Still loving goka.

One question has come up: if we're processing data that's almost in the right form for our View, but hasn't got the necessary key to   allow us to call ctx.SetValue(msg) we're having to make an intermediate topic whose only job is to receive a keyed message. 

```
group := goka.DefineGroup(group,
  goka.Input(topicWithoutKey, codec, AssignKey),
  goka.Output(topicWithKey, codec), 
  goka.Input(topicWithKey, codec, Store),
  goka.Persist(codec),
)
```

and then simply 
```
func AssignKey(ctx goka.Context, msg interface{}) {
        m := msg.(*SomeStruct)
        ctx.Emit(keyedtopic, m.Key, msg)
}

func Store(ctx goka.Context, msg interface{}) {
        ctx.SetValue(msg)
}
```
Feels clunky enough (though not the end of the world obviously) that it was worth checking if I'm missing something. `Loopback` still confuses me a bit but looks hopeful....

Am i missing something? Thanks in advance, and thanks as always for goka!

**Comment by mikedewar** [comment]
Oops turns out this has been asked before 

https://github.com/lovoo/goka/issues/123

Closing!

**Comment by frairon** [comment]
Hey @mikedewar,

actually there is a simpler way, like a shortcut, but it is quite unsafe and can introduce errors. 
The idea is, that If we have a processor, which only stores an object with the key it receives, we could also `Emit` that value directly to the target table, because that's what `SetValue` of the processor would merely do anyway.

Consider this group graph. It receives the message, loops it to the "correct key" and then simply stores it.
```go
goka.DefineGroup("group",
  goka.Input(input, codec, func(ctx goka.Context, msg interface{}){
       goka.Loopback("correct-key", msg)
}),
goka.Loop(codec, func(ctx goka.Context, msg Interface{}){
    ctx.SetValue(msg)
}),
goka.Persist(codec),
)

view, _ := goka.NewView("group-table", codec)
// we can use view to obtain the messages under the "correct" keys
```

What happens here is that the processor sends the message using the correct key back to itself via kafka (using the loopback topic), then stores it in its table. 

Advantages here:
* the processor is scaleable
* adding side effects is safe
* joining/lookup/aggregation on other tables is safe

On the other hand, assuming we read the values for the "correct keys" using a `View`, then using any topic as a source for the view is fine, it doesn't have to be a processor-table. The view consumes the updates and overwrites its local table.
Our assumption of course is, that the kafka topic backing the view has `cleanup-policy:compact`. Normally, the processor creates a topic for its own table with the correct configuration, but we can also create that manually like this:

```go

tmgr, _ := goka.NewTopicManager(...)
err := tmgr.EnsureTableExists("output-table", numPartitions)
// handle error
```

Now we can output directly into that topic using a processor:

```go

goka.DefineGroup("group",
goka.Input(input, codec, func(ctx goka.Context, msg interface{}){
    goka.Emit("output-table", "correct-key", msg)
}),
goka.Output("output-table", codec),
)

view, _ := goka.NewView("output-table", codec)
// we can use view to obtain the messages under the "correct" keys

```

Here we don't use an extra message via the topic, but emit directly into the target table. That emit will overwrite the existing value. But overwriting is all you can do. Suppose one day you want to aggregate the values in the table, or you want to modify it using a `lookup`-value, you would either risk race conditions or have to rewrite it and end up with the first form using loopback.

To summarize, you can avoid the extra processor, if you must - for performance reasons or if there is a good chance, you'll never need operations other than `Set`.

We use this approach e.g. to maintain simple kafka-backed maps in goka, where all we need to do is push an updated value and don't want to run an extra instance for the processor. If you want, I can give you more details on that approach, but maybe choose the other one from the beginning :) 

Hope that helped a bit.
